Jenkinsfile

```diff
@@ -1,28 +1,110 @@
 #!groovy
 library "atomSharedLibraries@master"
 
+import static Constants.*
+
+class Constants {
+    static final mainDirectoryName = "contrail_state_files"
+    static final terraformStateFileName = "state"
+}
+
+def warningEcho(message) {
+    echo "\033[1;33m${message}\033[0m"
+}
+
+def ensureMainDirectoryExists() {
+    dirExistingResult = checkIfDirectoryExists("../${mainDirectoryName}")
+    if(dirExistingResult == '0') {
+        sh "mkdir ../${mainDirectoryName}"
+    }
+}
+
+def ensureUserDirectoryExists() {
+    dirExistingResult = checkIfDirectoryExists("../${mainDirectoryName}/${params.Login}")
+    if(dirExistingResult == '0') {
+        sh "mkdir ../${mainDirectoryName}/${params.Login}"
+    }
+}
+
+def checkIfDirectoryExists(name) {
+    return sh(script: "test -d \"${name}\" && echo '1' || echo '0' ", returnStdout: true).trim()
+}
+
+def checkIfFileExists(name) {
+    return sh(script: "test -f \"${name}\" && echo '1' || echo '0' ", returnStdout: true).trim()
+}
+
+def copyTerraformFiles() {
+    sh "cp provision/${params.orchestrator}/${params.orchestrator}.tf ../${mainDirectoryName}/${params.Login}/${params.MachineName}/${params.orchestrator}.tf"
+    sh "cp provision/${params.orchestrator}/variables.tf ../${mainDirectoryName}/${params.Login}/${params.MachineName}/variables.tf"
+}
+
+def copyDaemonFile() {
+    sh "cp provision/daemon.json ../${mainDirectoryName}/${params.Login}/${params.MachineName}/daemon.json"
+}
+
+def resolveInstancesYamlFile() {
+    if ("${instances_yaml}" != "") {
+        def instancesYaml = unstashParam "instances_yaml"
+        sh "mv ${instancesYaml} ../${mainDirectoryName}/${params.Login}/${params.MachineName}/template.yaml"
+    } else {
+        warningEcho("Didn't get instances yaml file. Using default file.")
+        sh "cp provision/template.yaml ../${mainDirectoryName}/${params.Login}/${params.MachineName}/template.yaml"
+    }
+}
+
+def prepareMachineDirectory() {
+    sh "mkdir ../${mainDirectoryName}/${params.Login}/${params.MachineName}"
```

Comment:
I think I would suggest using absolute paths

Reply:
The main directory exists between other job directories. Since Jenkins has a special variable WORKSPACE which contains the absolute path from current job it's easy to deny on it.